In [254]:
# Importing necessary packages
import numpy as np 
import polars as pl 
import pandas as pd

In [255]:
# Reading in the data
df = pl.read_csv(r"C:\Users\gbore\Downloads\MMA\ufc.csv")
df2 = pl.read_csv(r"C:\Users\gbore\Downloads\MMA\ufc_fight_results.csv")

# Exploratory data analysis

In [256]:
# Taking a high level look at the data using head
df.head(n=15)

shape: (15, 144)
┌─────────────┬─────────────┬────────────┬────────────┬───┬───────────┬────────────┬───────┬───────┐
│ R_fighter   ┆ B_fighter   ┆ Referee    ┆ date       ┆ … ┆ R_Reach_c ┆ R_Weight_l ┆ B_age ┆ R_age │
│ ---         ┆ ---         ┆ ---        ┆ ---        ┆   ┆ ms        ┆ bs         ┆ ---   ┆ ---   │
│ str         ┆ str         ┆ str        ┆ str        ┆   ┆ ---       ┆ ---        ┆ f64   ┆ f64   │
│             ┆             ┆            ┆            ┆   ┆ f64       ┆ f64        ┆       ┆       │
╞═════════════╪═════════════╪════════════╪════════════╪═══╪═══════════╪════════════╪═══════╪═══════╡
│ Adrian      ┆ Gustavo     ┆ Chris      ┆ 2021-03-20 ┆ … ┆ 177.8     ┆ 135.0      ┆ 31.0  ┆ 27.0  │
│ Yanez       ┆ Lopez       ┆ Tognoni    ┆            ┆   ┆           ┆            ┆       ┆       │
│ Trevin      ┆ Roman       ┆ Herb Dean  ┆ 2021-03-20 ┆ … ┆ 187.96    ┆ 185.0      ┆ 32.0  ┆ 28.0  │
│ Giles       ┆ Dolidze     ┆            ┆            ┆   ┆           ┆            ┆       ┆       │
│ Tai Tuivasa ┆ Harry       ┆ Herb Dean  ┆ 2021-03-20 ┆ … ┆ 190.5     ┆ 264.0      ┆ 32.0  ┆ 28.0  │
│             ┆ Hunsucker   ┆            ┆            ┆   ┆           ┆            ┆       ┆       │
│ Cheyanne    ┆ Montserrat  ┆ Mark Smith ┆ 2021-03-20 ┆ … ┆ 160.02    ┆ 115.0      ┆ 28.0  ┆ 25.0  │
│ Buys        ┆ Conejo      ┆            ┆            ┆   ┆           ┆            ┆       ┆       │
│ …           ┆ …           ┆ …          ┆ …          ┆ … ┆ …         ┆ …          ┆ …     ┆ …     │
│ Misha       ┆ Ryan Spann  ┆ Mark Smith ┆ 2021-03-13 ┆ … ┆ 195.58    ┆ 205.0      ┆ 29.0  ┆ 34.0  │
│ Cirkunov    ┆             ┆            ┆            ┆   ┆           ┆            ┆       ┆       │
│ Leon        ┆ Belal       ┆ Herb Dean  ┆ 2021-03-13 ┆ … ┆ 187.96    ┆ 170.0      ┆ 32.0  ┆ 29.0  │
│ Edwards     ┆ Muhammad    ┆            ┆            ┆   ┆           ┆            ┆       ┆       │
│ Charles     ┆ Marcelo     ┆ Mark Smith ┆ 2021-03-13 ┆ … ┆ 175.26    ┆ 145.0      ┆ 32.0  ┆ 25.0  │
│ Jourdain    ┆ Rojo        ┆            ┆            ┆   ┆           ┆            ┆       ┆       │
│ Eryk Anders ┆ Darren      ┆ Herb Dean  ┆ 2021-03-13 ┆ … ┆ 190.5     ┆ 185.0      ┆ 30.0  ┆ 33.0  │
│             ┆ Stewart     ┆            ┆            ┆   ┆           ┆            ┆       ┆       │
└─────────────┴─────────────┴────────────┴────────────┴───┴───────────┴────────────┴───────┴───────┘

We looked at the first 5 rows(or 5 latest fights) of each of the 144 columns here and in the next line of code.

In [257]:
# Taking a high level look at the data using head
df2.head(n=15)

shape: (15, 11)
┌────────────┬───────────┬─────────┬───────────┬───┬───────────┬───────────┬───────────┬───────────┐
│ EVENT      ┆ BOUT      ┆ OUTCOME ┆ WEIGHTCLA ┆ … ┆ TIME      ┆ REFEREE   ┆ DETAILS   ┆ URL       │
│ ---        ┆ ---       ┆ ---     ┆ SS        ┆   ┆ FORMAT    ┆ ---       ┆ ---       ┆ ---       │
│ str        ┆ str       ┆ str     ┆ ---       ┆   ┆ ---       ┆ str       ┆ str       ┆ str       │
│            ┆           ┆         ┆ str       ┆   ┆ str       ┆           ┆           ┆           │
╞════════════╪═══════════╪═════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪═══════════╡
│ UFC Fight  ┆ Derek     ┆ W/L     ┆ Middlewei ┆ … ┆ 5 Rnd (5- ┆ Herb Dean ┆ Mike Bell ┆ http://uf │
│ Night:     ┆ Brunson   ┆         ┆ ght Bout  ┆   ┆ 5-5-5-5)  ┆           ┆ 45 -      ┆ cstats.co │
│ Brunson    ┆ vs. Kevin ┆         ┆           ┆   ┆           ┆           ┆ 49.Sal    ┆ m/fight-d │
│ vs. Hol…   ┆ Holland…  ┆         ┆           ┆   ┆           ┆           ┆ D'amato   ┆ etail…    │
│            ┆           ┆         ┆           ┆   ┆           ┆           ┆ 46…       ┆           │
│ UFC Fight  ┆ Song      ┆ L/W     ┆ Welterwei ┆ … ┆ 3 Rnd     ┆ Keith     ┆ Punch to  ┆ http://uf │
│ Night:     ┆ Kenan     ┆         ┆ ght Bout  ┆   ┆ (5-5-5)   ┆ Peterson  ┆ Head At   ┆ cstats.co │
│ Brunson    ┆ vs. Max   ┆         ┆           ┆   ┆           ┆           ┆ Distance  ┆ m/fight-d │
│ vs. Hol…   ┆ Griffin   ┆         ┆           ┆   ┆           ┆           ┆           ┆ etail…    │
│ UFC Fight  ┆ Cheyanne  ┆ L/W     ┆ Women's   ┆ … ┆ 3 Rnd     ┆ Mark      ┆ Mike Bell ┆ http://uf │
│ Night:     ┆ Vlismas   ┆         ┆ Strawweig ┆   ┆ (5-5-5)   ┆ Smith     ┆ 28 - 29.J ┆ cstats.co │
│ Brunson    ┆ vs. Monts ┆         ┆ ht Bout   ┆   ┆           ┆           ┆ unichiro  ┆ m/fight-d │
│ vs. Hol…   ┆ errat…    ┆         ┆           ┆   ┆           ┆           ┆ Kami…     ┆ etail…    │
│ UFC Fight  ┆ Adrian    ┆ W/L     ┆ Bantamwei ┆ … ┆ 3 Rnd     ┆ Chris     ┆ Punch to  ┆ http://uf │
│ Night:     ┆ Yanez     ┆         ┆ ght Bout  ┆   ┆ (5-5-5)   ┆ Tognoni   ┆ Head At   ┆ cstats.co │
│ Brunson    ┆ vs.       ┆         ┆           ┆   ┆           ┆           ┆ Distance  ┆ m/fight-d │
│ vs. Hol…   ┆ Gustavo   ┆         ┆           ┆   ┆           ┆           ┆           ┆ etail…    │
│            ┆ Lopez     ┆         ┆           ┆   ┆           ┆           ┆           ┆           │
│ …          ┆ …         ┆ …       ┆ …         ┆ … ┆ …         ┆ …         ┆ …         ┆ …         │
│ UFC Fight  ┆ Misha     ┆ L/W     ┆ Light Hea ┆ … ┆ 3 Rnd     ┆ Mark      ┆ Punches   ┆ http://uf │
│ Night:     ┆ Cirkunov  ┆         ┆ vyweight  ┆   ┆ (5-5-5)   ┆ Smith     ┆ to Head   ┆ cstats.co │
│ Edwards    ┆ vs. Ryan  ┆         ┆ Bout      ┆   ┆           ┆           ┆ On Ground ┆ m/fight-d │
│ vs. Muh…   ┆ Spann     ┆         ┆           ┆   ┆           ┆           ┆           ┆ etail…    │
│ UFC Fight  ┆ Dan Ige   ┆ W/L     ┆ Featherwe ┆ … ┆ 3 Rnd     ┆ Mike      ┆ Punch to  ┆ http://uf │
│ Night:     ┆ vs. Gavin ┆         ┆ ight Bout ┆   ┆ (5-5-5)   ┆ Beltran   ┆ Head At   ┆ cstats.co │
│ Edwards    ┆ Tucker    ┆         ┆           ┆   ┆           ┆           ┆ Distance  ┆ m/fight-d │
│ vs. Muh…   ┆           ┆         ┆           ┆   ┆           ┆           ┆           ┆ etail…    │
│ UFC Fight  ┆ Jonathan  ┆ L/W     ┆ Bantamwei ┆ … ┆ 3 Rnd     ┆ Chris     ┆ Punch to  ┆ http://uf │
│ Night:     ┆ Martinez  ┆         ┆ ght Bout  ┆   ┆ (5-5-5)   ┆ Tognoni   ┆ Head At   ┆ cstats.co │
│ Edwards    ┆ vs. Davey ┆         ┆           ┆   ┆           ┆           ┆ Distance  ┆ m/fight-d │
│ vs. Muh…   ┆ Gra…      ┆         ┆           ┆   ┆           ┆           ┆           ┆ etail…    │
│ UFC Fight  ┆ Manel     ┆ L/W     ┆ Flyweight ┆ … ┆ 3 Rnd     ┆ null      ┆ Chris Lee ┆ http://uf │
│ Night:     ┆ Kape  vs. ┆         ┆ Bout      ┆   ┆ (5-5-5)   ┆           ┆ 28 -      ┆ cstats.co │
│ Edwards    ┆ Matheus   ┆         ┆         

In [258]:
# Taking a high level look at the data using tail
df.tail(n=15)

shape: (15, 144)
┌─────────────┬─────────────┬────────────┬────────────┬───┬───────────┬────────────┬───────┬───────┐
│ R_fighter   ┆ B_fighter   ┆ Referee    ┆ date       ┆ … ┆ R_Reach_c ┆ R_Weight_l ┆ B_age ┆ R_age │
│ ---         ┆ ---         ┆ ---        ┆ ---        ┆   ┆ ms        ┆ bs         ┆ ---   ┆ ---   │
│ str         ┆ str         ┆ str        ┆ str        ┆   ┆ ---       ┆ ---        ┆ f64   ┆ f64   │
│             ┆             ┆            ┆            ┆   ┆ f64       ┆ f64        ┆       ┆       │
╞═════════════╪═════════════╪════════════╪════════════╪═══╪═══════════╪════════════╪═══════╪═══════╡
│ Royce       ┆ Patrick     ┆ John       ┆ 1994-03-11 ┆ … ┆ null      ┆ 175.0      ┆ 30.0  ┆ 27.0  │
│ Gracie      ┆ Smith       ┆ McCarthy   ┆            ┆   ┆           ┆            ┆       ┆       │
│ Patrick     ┆ Scott       ┆ John       ┆ 1994-03-11 ┆ … ┆ null      ┆ 225.0      ┆ null  ┆ 30.0  │
│ Smith       ┆ Morris      ┆ McCarthy   ┆            ┆   ┆           ┆            ┆       ┆       │
│ Royce       ┆ Remco       ┆ John       ┆ 1994-03-11 ┆ … ┆ null      ┆ 175.0      ┆ null  ┆ 27.0  │
│ Gracie      ┆ Pardoel     ┆ McCarthy   ┆            ┆   ┆           ┆            ┆       ┆       │
│ Remco       ┆ Orlando     ┆ John       ┆ 1994-03-11 ┆ … ┆ null      ┆ 260.0      ┆ null  ┆ null  │
│ Pardoel     ┆ Wiet        ┆ McCarthy   ┆            ┆   ┆           ┆            ┆       ┆       │
│ …           ┆ …           ┆ …          ┆ …          ┆ … ┆ …         ┆ …          ┆ …     ┆ …     │
│ Orlando     ┆ Robert      ┆ John       ┆ 1994-03-11 ┆ … ┆ null      ┆ 170.0      ┆ null  ┆ null  │
│ Wiet        ┆ Lucarelli   ┆ McCarthy   ┆            ┆   ┆           ┆            ┆       ┆       │
│ Johnny      ┆ David       ┆ John       ┆ 1994-03-11 ┆ … ┆ null      ┆ 210.0      ┆ null  ┆ null  │
│ Rhodes      ┆ Levicki     ┆ McCarthy   ┆            ┆   ┆           ┆            ┆       ┆       │
│ Patrick     ┆ Ray Wizard  ┆ John       ┆ 1994-03-11 ┆ … ┆ null      ┆ 225.0      ┆ null  ┆ 30.0  │
│ Smith       ┆             ┆ McCarthy   ┆            ┆   ┆           ┆            ┆       ┆       │
│ Scott       ┆ Sean        ┆ John       ┆ 1994-03-11 ┆ … ┆ null      ┆ 210.0      ┆ 18.0  ┆ null  │
│ Morris      ┆ Daugherty   ┆ McCarthy   ┆            ┆   ┆           ┆            ┆       ┆       │
└─────────────┴─────────────┴────────────┴────────────┴───┴───────────┴────────────┴───────┴───────┘

We looked at the last 5 rows(or first 5 fights) of each of the 144 columns here and in the next line of code.

In [259]:
df2.tail(n=15)

shape: (15, 11)
┌────────────┬───────────┬─────────┬───────────┬───┬───────────┬───────────┬───────────┬───────────┐
│ EVENT      ┆ BOUT      ┆ OUTCOME ┆ WEIGHTCLA ┆ … ┆ TIME      ┆ REFEREE   ┆ DETAILS   ┆ URL       │
│ ---        ┆ ---       ┆ ---     ┆ SS        ┆   ┆ FORMAT    ┆ ---       ┆ ---       ┆ ---       │
│ str        ┆ str       ┆ str     ┆ ---       ┆   ┆ ---       ┆ str       ┆ str       ┆ str       │
│            ┆           ┆         ┆ str       ┆   ┆ str       ┆           ┆           ┆           │
╞════════════╪═══════════╪═════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪═══════════╡
│ UFC 2: No  ┆ Royce     ┆ W/L     ┆ UFC 2 Tou ┆ … ┆ No Time   ┆ John      ┆ Punches   ┆ http://uf │
│ Way Out    ┆ Gracie    ┆         ┆ rnament   ┆   ┆ Limit     ┆ McCarthy  ┆ to Head   ┆ cstats.co │
│            ┆ vs.       ┆         ┆ Title     ┆   ┆           ┆           ┆ From Moun ┆ m/fight-d │
│            ┆ Patrick   ┆         ┆ Bout      ┆   ┆           ┆           ┆ tSubmis…  ┆ etail…    │
│            ┆ Smith     ┆         ┆           ┆   ┆           ┆           ┆           ┆           │
│ UFC 2: No  ┆ Royce     ┆ W/L     ┆ Open      ┆ … ┆ No Time   ┆ John      ┆ Gi Choke  ┆ http://uf │
│ Way Out    ┆ Gracie    ┆         ┆ Weight    ┆   ┆ Limit     ┆ McCarthy  ┆ From Back ┆ cstats.co │
│            ┆ vs. Remco ┆         ┆ Bout      ┆   ┆           ┆           ┆ Control   ┆ m/fight-d │
│            ┆ Pardoel   ┆         ┆           ┆   ┆           ┆           ┆           ┆ etail…    │
│ UFC 2: No  ┆ Patrick   ┆ W/L     ┆ Open      ┆ … ┆ No Time   ┆ John      ┆ Guillotin ┆ http://uf │
│ Way Out    ┆ Smith     ┆         ┆ Weight    ┆   ┆ Limit     ┆ McCarthy  ┆ e Choke   ┆ cstats.co │
│            ┆ vs.       ┆         ┆ Bout      ┆   ┆           ┆           ┆ Standing  ┆ m/fight-d │
│            ┆ Johnny    ┆         ┆           ┆   ┆           ┆           ┆           ┆ etail…    │
│            ┆ Rhodes…   ┆         ┆           ┆   ┆           ┆           ┆           ┆           │
│ UFC 2: No  ┆ Royce     ┆ W/L     ┆ Open      ┆ … ┆ No Time   ┆ John      ┆ Armbar    ┆ http://uf │
│ Way Out    ┆ Gracie    ┆         ┆ Weight    ┆   ┆ Limit     ┆ McCarthy  ┆ From      ┆ cstats.co │
│            ┆ vs. Jason ┆         ┆ Bout      ┆   ┆           ┆           ┆ Mount     ┆ m/fight-d │
│            ┆ DeLucia   ┆         ┆           ┆   ┆           ┆           ┆           ┆ etail…    │
│ …          ┆ …         ┆ …       ┆ …         ┆ … ┆ …         ┆ …         ┆ …         ┆ …         │
│ UFC 2: No  ┆ Frank     ┆ W/L     ┆ Open      ┆ … ┆ No Time   ┆ John      ┆ Keylock   ┆ http://uf │
│ Way Out    ┆ Hamaker   ┆         ┆ Weight    ┆   ┆ Limit     ┆ McCarthy  ┆ From Half ┆ cstats.co │
│            ┆ vs.       ┆         ┆ Bout      ┆   ┆           ┆           ┆ Guard     ┆ m/fight-d │
│            ┆ Thaddeus  ┆         ┆           ┆   ┆           ┆           ┆           ┆ etail…    │
│            ┆ Lust…     ┆         ┆           ┆   ┆           ┆           ┆           ┆           │
│ UFC 2: No  ┆ Johnny    ┆ W/L     ┆ Open      ┆ … ┆ No Time   ┆ John      ┆ Punches   ┆ http://uf │
│ Way Out    ┆ Rhodes    ┆         ┆ Weight    ┆   ┆ Limit     ┆ McCarthy  ┆ to Head   ┆ cstats.co │
│            ┆ vs. David ┆         ┆ Bout      ┆   ┆           ┆           ┆ From Guar ┆ m/fight-d │
│            ┆ Levicki…  ┆         ┆           ┆   ┆           ┆           ┆ dSubmis…  ┆ etail…    │
│ UFC 2: No  ┆ Patrick   ┆ W/L     ┆ Open      ┆ … ┆ No Time   ┆ John      ┆ Guillotin ┆ http://uf │
│ Way Out    ┆ Smith     ┆         ┆ Weight    ┆   ┆ Limit     ┆ McCarthy  ┆ e Choke   ┆ cstats.co │
│            ┆ vs. Ray   ┆         ┆ Bout      ┆   ┆           ┆           ┆ Standing  ┆ m/fight-d │
│            ┆ Wizard    ┆         ┆           ┆   ┆           ┆           ┆           ┆ etail…    │
│ UFC 2: No  ┆ Scott     ┆ W/L     ┆ Open      ┆ … ┆ No Time   ┆ John      ┆ Guillotin ┆ http://uf │
│ Way Out    ┆ Morris    ┆         ┆ Weight  

In [260]:
# Looking at and understanding the names/meanings of the columns
df.columns

['R_fighter',
 'B_fighter',
 'Referee',
 'date',
 'location',
 'Winner',
 'title_bout',
 'weight_class',
 'B_avg_KD',
 'B_avg_opp_KD',
 'B_avg_SIG_STR_pct',
 'B_avg_opp_SIG_STR_pct',
 'B_avg_TD_pct',
 'B_avg_opp_TD_pct',
 'B_avg_SUB_ATT',
 'B_avg_opp_SUB_ATT',
 'B_avg_REV',
 'B_avg_opp_REV',
 'B_avg_SIG_STR_att',
 'B_avg_SIG_STR_landed',
 'B_avg_opp_SIG_STR_att',
 'B_avg_opp_SIG_STR_landed',
 'B_avg_TOTAL_STR_att',
 'B_avg_TOTAL_STR_landed',
 'B_avg_opp_TOTAL_STR_att',
 'B_avg_opp_TOTAL_STR_landed',
 'B_avg_TD_att',
 'B_avg_TD_landed',
 'B_avg_opp_TD_att',
 'B_avg_opp_TD_landed',
 'B_avg_HEAD_att',
 'B_avg_HEAD_landed',
 'B_avg_opp_HEAD_att',
 'B_avg_opp_HEAD_landed',
 'B_avg_BODY_att',
 'B_avg_BODY_landed',
 'B_avg_opp_BODY_att',
 'B_avg_opp_BODY_landed',
 'B_avg_LEG_att',
 'B_avg_LEG_landed',
 'B_avg_opp_LEG_att',
 'B_avg_opp_LEG_landed',
 'B_avg_DISTANCE_att',
 'B_avg_DISTANCE_landed',
 'B_avg_opp_DISTANCE_att',
 'B_avg_opp_DISTANCE_landed',
 'B_avg_CLINCH_att',
 'B_avg_CLINCH_lande

The column names are shortened but record 
- Fighter Names(Blue and Red Corner)
- Generic information: Date, location, referee name, whether the fight was for a title and the weight class
- Blue Fighter average fighting career statistics for/against: knockdowns, significant strikes percentage/attempted/landed, takedown percentage/attempted/landed, submissions, reversals, total strikes attempted/landed, head/body/leg strikes attempted/landed, attacks landed/attempted at clinch/distance/ground ranges, control time for/against
- Blue fighter generic career statistics: Total time/rounds fought, title bouts, win/lose streak, longest winning streak, total wins/losses/draws, wins by majority/split/unanimous decision, wins by ko/tko, wins by submission, wins by doctor stoppage
- Blue fighter physiological statistics: Stance, Height(cms), Weight, Reach
- Red Fighter average fighting career statistics for/against: knockdowns, significant strikes percentage/attempted/landed, takedown percentage/attempted/landed, submissions, reversals, total strikes attempted/landed, head/body/leg strikes attempted/landed, attacks landed/attempted at clinch/distance/ground ranges, control time for/against
- Red fighter generic career statistics: Total time/rounds fought, title bouts, win/lose streak, longest winning streak, total wins/losses/draws, wins by majority/split/unanimous decision, wins by ko/tko, wins by submission, wins by doctor stoppage
- Red fighter physiological statistics: Stance, Height(cms), Weight, Reach

Since this is a fairly new sport, some non-obvious but important terms need to be clarified for ease of understanding to non-fight fans, here are the definitions:-

- Knockout (KO): When an athlete is knocked unconscious due to strikes or other impact. A knockout is not a technical knockout, but rather one that leaves the fighter physically unable to continue.
- Technical Knockout (TKO): Abbreviation for technical knockout, when an athlete is unable to continue, usually due to injury.
- Submission: When an athlete taps out or verbally concedes the match due to pain, to avoid injury, to avoid being choked or due to a desire to end the match. A referee has the power to force a fighter to submit at his discretion due to risk of severe injury/loss of consciousness/death.
- Knockdown (KD): When a fighter touches the floor of the ring with any part of the body other than the feet following a hit, but is able to rise back up and continue fighting
- Reversal: When an athlete moves from an inferior position into a superior position.
- Unanimous Decision: All judges scored the fight in favor of one fighter.
- Majority Decision: 2 judges scored the fight in favor of the winner while the third scored a draw.
- Split Decision: 2 judges scored one corner as the winner while the third voted for the opposing corner.
- Majority decision draw: Two judges voted for a draw while one judge voted for a winner.
- Split decision draw: The three judges scored the contest as blue corner win, red corner win and draw.

In [261]:
#Counting the number of times a value appears in each column
for x in df:
    y = x.value_counts().sort(by = 'counts', descending = True)
    print(y)

shape: (1_514, 2)
┌─────────────────┬────────┐
│ R_fighter       ┆ counts │
│ ---             ┆ ---    │
│ str             ┆ u32    │
╞═════════════════╪════════╡
│ Jim Miller      ┆ 27     │
│ Andrei Arlovski ┆ 24     │
│ Donald Cerrone  ┆ 24     │
│ Diego Sanchez   ┆ 24     │
│ …               ┆ …      │
│ Gabriel Silva   ┆ 1      │
│ Alexis Dufresne ┆ 1      │
│ Iliarde Santos  ┆ 1      │
│ Luiz Dutra      ┆ 1      │
└─────────────────┴────────┘
shape: (1_987, 2)
┌───────────────────────┬────────┐
│ B_fighter             ┆ counts │
│ ---                   ┆ ---    │
│ str                   ┆ u32    │
╞═══════════════════════╪════════╡
│ Jeremy Stephens       ┆ 21     │
│ Charles Oliveira      ┆ 19     │
│ Nik Lentz             ┆ 15     │
│ Rafael Dos Anjos      ┆ 14     │
│ …                     ┆ …      │
│ Andre Muniz           ┆ 1      │
│ Anthony Christodoulou ┆ 1      │
│ John Marsh            ┆ 1      │
│ Ludovit Klein         ┆ 1      │
└───────────────────────┴────────┘
shap

In [262]:
#Taking a look at the shape of the data
df.shape

(6012, 144)

This reveals there have been 6012 fights and 144 fighter statistics have been tracked in the columns. The column names and what they mean have been described above.

In [263]:
# Reading in the data using pandas
dfpandas = pd.read_csv(r"C:\Users\gbore\Downloads\MMA\ufc.csv")
df2pandas = pd.read_csv(r"C:\Users\gbore\Downloads\MMA\ufc_fight_results.csv")

In [264]:
# Looking at the data structures of the dataframes
dfpandas.info()
df2pandas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6012 entries, 0 to 6011
Columns: 144 entries, R_fighter to R_age
dtypes: bool(1), float64(106), int64(28), object(9)
memory usage: 6.6+ MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6033 entries, 0 to 6032
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   EVENT        6033 non-null   object
 1   BOUT         6033 non-null   object
 2   OUTCOME      6033 non-null   object
 3   WEIGHTCLASS  6033 non-null   object
 4   METHOD       6033 non-null   object
 5   ROUND        6033 non-null   int64 
 6   TIME         6033 non-null   object
 7   TIME FORMAT  6033 non-null   object
 8   REFEREE      6001 non-null   object
 9   DETAILS      5962 non-null   object
 10  URL          6033 non-null   object
dtypes: int64(1), object(10)
memory usage: 518.6+ KB


In [265]:
# Printing the names and types of each feature
for x in dfpandas:
    print("Name(Type) : " + x + "(" + str(type(x)) + ")")

Name(Type) : R_fighter(<class 'str'>)
Name(Type) : B_fighter(<class 'str'>)
Name(Type) : Referee(<class 'str'>)
Name(Type) : date(<class 'str'>)
Name(Type) : location(<class 'str'>)
Name(Type) : Winner(<class 'str'>)
Name(Type) : title_bout(<class 'str'>)
Name(Type) : weight_class(<class 'str'>)
Name(Type) : B_avg_KD(<class 'str'>)
Name(Type) : B_avg_opp_KD(<class 'str'>)
Name(Type) : B_avg_SIG_STR_pct(<class 'str'>)
Name(Type) : B_avg_opp_SIG_STR_pct(<class 'str'>)
Name(Type) : B_avg_TD_pct(<class 'str'>)
Name(Type) : B_avg_opp_TD_pct(<class 'str'>)
Name(Type) : B_avg_SUB_ATT(<class 'str'>)
Name(Type) : B_avg_opp_SUB_ATT(<class 'str'>)
Name(Type) : B_avg_REV(<class 'str'>)
Name(Type) : B_avg_opp_REV(<class 'str'>)
Name(Type) : B_avg_SIG_STR_att(<class 'str'>)
Name(Type) : B_avg_SIG_STR_landed(<class 'str'>)
Name(Type) : B_avg_opp_SIG_STR_att(<class 'str'>)
Name(Type) : B_avg_opp_SIG_STR_landed(<class 'str'>)
Name(Type) : B_avg_TOTAL_STR_att(<class 'str'>)
Name(Type) : B_avg_TOTAL_STR_

In [266]:
# Repeating with fight statistics
for x in df2pandas:
    print("Name(Type) : " + x + "(" + str(type(x)) + ")")

Name(Type) : EVENT(<class 'str'>)
Name(Type) : BOUT(<class 'str'>)
Name(Type) : OUTCOME(<class 'str'>)
Name(Type) : WEIGHTCLASS(<class 'str'>)
Name(Type) : METHOD(<class 'str'>)
Name(Type) : ROUND(<class 'str'>)
Name(Type) : TIME(<class 'str'>)
Name(Type) : TIME FORMAT(<class 'str'>)
Name(Type) : REFEREE(<class 'str'>)
Name(Type) : DETAILS(<class 'str'>)
Name(Type) : URL(<class 'str'>)


In [267]:
# Looking at summary statistics for the fighter's careers
df.describe()

shape: (9, 145)
┌───────────┬─────────┬───────────┬───────────┬───┬───────────┬────────────┬───────────┬───────────┐
│ describe  ┆ R_fight ┆ B_fighter ┆ Referee   ┆ … ┆ R_Reach_c ┆ R_Weight_l ┆ B_age     ┆ R_age     │
│ ---       ┆ er      ┆ ---       ┆ ---       ┆   ┆ ms        ┆ bs         ┆ ---       ┆ ---       │
│ str       ┆ ---     ┆ str       ┆ str       ┆   ┆ ---       ┆ ---        ┆ f64       ┆ f64       │
│           ┆ str     ┆           ┆           ┆   ┆ f64       ┆ f64        ┆           ┆           │
╞═══════════╪═════════╪═══════════╪═══════════╪═══╪═══════════╪════════════╪═══════════╪═══════════╡
│ count     ┆ 6012    ┆ 6012      ┆ 6012      ┆ … ┆ 6012.0    ┆ 6012.0     ┆ 6012.0    ┆ 6012.0    │
│ null_coun ┆ 0       ┆ 0         ┆ 32        ┆ … ┆ 406.0     ┆ 2.0        ┆ 172.0     ┆ 63.0      │
│ t         ┆         ┆           ┆           ┆   ┆           ┆            ┆           ┆           │
│ mean      ┆ null    ┆ null      ┆ null      ┆ … ┆ 183.56234 ┆ 170.686356 ┆ 29.293151 ┆ 29.598252 │
│           ┆         ┆           ┆           ┆   ┆ 7         ┆            ┆           ┆           │
│ std       ┆ null    ┆ null      ┆ null      ┆ … ┆ 10.585313 ┆ 35.724626  ┆ 4.063297  ┆ 4.167755  │
│ min       ┆ Aalon   ┆ Aalon     ┆ Adam      ┆ … ┆ 152.4     ┆ 115.0      ┆ 18.0      ┆ 19.0      │
│           ┆ Cruz    ┆ Cruz      ┆ Cheadle   ┆   ┆           ┆            ┆           ┆           │
│ max       ┆ Zubaira ┆ Zubaira   ┆ Yves      ┆ … ┆ 213.36    ┆ 345.0      ┆ 51.0      ┆ 47.0      │
│           ┆ Tukhugo ┆ Tukhugov  ┆ Lavigne   ┆   ┆           ┆            ┆           ┆           │
│           ┆ v       ┆           ┆           ┆   ┆           ┆            ┆           ┆           │
│ median    ┆ null    ┆ null      ┆ null      ┆ … ┆ 182.88    ┆ 170.0      ┆ 29.0      ┆ 29.0      │
│ 25%       ┆ null    ┆ null      ┆ null      ┆ … ┆ 177.8     ┆ 145.0      ┆ 26.0      ┆ 27.0      │
│ 75%       ┆ null    ┆ null      ┆ null      ┆ … ┆ 190.5     ┆ 185.0      ┆ 32.0      ┆ 32.0      │
└───────────┴─────────┴───────────┴───────────┴───┴───────────┴────────────┴───────────┴───────────┘

In [268]:
# Looking at fight statistics 
df2.describe()

shape: (9, 12)
┌────────────┬───────────┬───────────┬─────────┬───┬───────────┬───────────┬───────────┬───────────┐
│ describe   ┆ EVENT     ┆ BOUT      ┆ OUTCOME ┆ … ┆ TIME      ┆ REFEREE   ┆ DETAILS   ┆ URL       │
│ ---        ┆ ---       ┆ ---       ┆ ---     ┆   ┆ FORMAT    ┆ ---       ┆ ---       ┆ ---       │
│ str        ┆ str       ┆ str       ┆ str     ┆   ┆ ---       ┆ str       ┆ str       ┆ str       │
│            ┆           ┆           ┆         ┆   ┆ str       ┆           ┆           ┆           │
╞════════════╪═══════════╪═══════════╪═════════╪═══╪═══════════╪═══════════╪═══════════╪═══════════╡
│ count      ┆ 6033      ┆ 6033      ┆ 6033    ┆ … ┆ 6033      ┆ 6033      ┆ 6033      ┆ 6033      │
│ null_count ┆ 0         ┆ 0         ┆ 0       ┆ … ┆ 0         ┆ 32        ┆ 71        ┆ 0         │
│ mean       ┆ null      ┆ null      ┆ null    ┆ … ┆ null      ┆ null      ┆ null      ┆ null      │
│ std        ┆ null      ┆ null      ┆ null    ┆ … ┆ null      ┆ null      ┆ null      ┆ null      │
│ min        ┆ Ortiz vs  ┆ Alatenghe ┆ D/D     ┆ … ┆ 1 Rnd     ┆ Adam      ┆ 26 -      ┆ http://uf │
│            ┆ Shamrock  ┆ ili  vs.  ┆         ┆   ┆ (10)      ┆ Cheadle   ┆ 30.26 -   ┆ cstats.co │
│            ┆ 3: The    ┆ Danaa     ┆         ┆   ┆           ┆           ┆ 30.27 -   ┆ m/fight-d │
│            ┆ Final C…  ┆ Batgere…  ┆         ┆   ┆           ┆           ┆ 29.       ┆ etail…    │
│ max        ┆ UFC:      ┆ Zubaira   ┆ W/L     ┆ … ┆ No Time   ┆ Yves      ┆ toVerbal  ┆ http://uf │
│            ┆ Silva vs  ┆ Tukhugov  ┆         ┆   ┆ Limit     ┆ Lavigne   ┆ Submissio ┆ cstats.co │
│            ┆ Irvin     ┆ vs.       ┆         ┆   ┆           ┆           ┆ n - Arm   ┆ m/fight-d │
│            ┆           ┆ Phillipe  ┆         ┆   ┆           ┆           ┆ Injury    ┆ etail…    │
│            ┆           ┆ N…        ┆         ┆   ┆           ┆           ┆           ┆           │
│ median     ┆ null      ┆ null      ┆ null    ┆ … ┆ null      ┆ null      ┆ null      ┆ null      │
│ 25%        ┆ null      ┆ null      ┆ null    ┆ … ┆ null      ┆ null      ┆ null      ┆ null      │
│ 75%        ┆ null      ┆ null      ┆ null    ┆ … ┆ null      ┆ null      ┆ null      ┆ null      │
└────────────┴───────────┴───────────┴─────────┴───┴───────────┴───────────┴───────────┴───────────┘

In [269]:
dfpandas.duplicated().sum()

0

In [270]:
df2pandas.duplicated().sum()

0

In [271]:
help(dfpandas.duplicated)

Help on method duplicated in module pandas.core.frame:

duplicated(subset: 'Hashable | Sequence[Hashable] | None' = None, keep: "Literal['first', 'last', False]" = 'first') -> 'Series' method of pandas.core.frame.DataFrame instance
    Return boolean Series denoting duplicate rows.
    
    Considering certain columns is optional.
    
    Parameters
    ----------
    subset : column label or sequence of labels, optional
        Only consider certain columns for identifying duplicates, by
        default use all of the columns.
    keep : {'first', 'last', False}, default 'first'
        Determines which duplicates (if any) to mark.
    
        - ``first`` : Mark duplicates as ``True`` except for the first occurrence.
        - ``last`` : Mark duplicates as ``True`` except for the last occurrence.
        - False : Mark all duplicates as ``True``.
    
    Returns
    -------
    Series
        Boolean series for each duplicated rows.
    
    See Also
    --------
    Index.duplicat

In [272]:
# Creating a dataframe to look at whether there are null values and how many
dfnull = pd.DataFrame(dfpandas.isnull().sum())
dfnull = dfnull.rename(columns = {0:'count'})

In [273]:
# Looking at what the indices and their values are
indices = dfnull.index
values = list(dfnull.values)

They seem to be just the features. Thus, we can create a dictionary counting the number of null values for each feature.

In [274]:
nullvaluedict = dict(zip(indices, values))

In [275]:
nullvaluedict

{'R_fighter': array([0], dtype=int64),
 'B_fighter': array([0], dtype=int64),
 'Referee': array([32], dtype=int64),
 'date': array([0], dtype=int64),
 'location': array([0], dtype=int64),
 'Winner': array([0], dtype=int64),
 'title_bout': array([0], dtype=int64),
 'weight_class': array([0], dtype=int64),
 'B_avg_KD': array([1427], dtype=int64),
 'B_avg_opp_KD': array([1427], dtype=int64),
 'B_avg_SIG_STR_pct': array([1427], dtype=int64),
 'B_avg_opp_SIG_STR_pct': array([1427], dtype=int64),
 'B_avg_TD_pct': array([1427], dtype=int64),
 'B_avg_opp_TD_pct': array([1427], dtype=int64),
 'B_avg_SUB_ATT': array([1427], dtype=int64),
 'B_avg_opp_SUB_ATT': array([1427], dtype=int64),
 'B_avg_REV': array([1427], dtype=int64),
 'B_avg_opp_REV': array([1427], dtype=int64),
 'B_avg_SIG_STR_att': array([1427], dtype=int64),
 'B_avg_SIG_STR_landed': array([1427], dtype=int64),
 'B_avg_opp_SIG_STR_att': array([1427], dtype=int64),
 'B_avg_opp_SIG_STR_landed': array([1427], dtype=int64),
 'B_avg_TOTA

In [276]:
z = []
for (x,y) in zip(nullvaluedict.values(), nullvaluedict):
    if x != 0:
        z.append(y)
print(z)
print(len(z))

['Referee', 'B_avg_KD', 'B_avg_opp_KD', 'B_avg_SIG_STR_pct', 'B_avg_opp_SIG_STR_pct', 'B_avg_TD_pct', 'B_avg_opp_TD_pct', 'B_avg_SUB_ATT', 'B_avg_opp_SUB_ATT', 'B_avg_REV', 'B_avg_opp_REV', 'B_avg_SIG_STR_att', 'B_avg_SIG_STR_landed', 'B_avg_opp_SIG_STR_att', 'B_avg_opp_SIG_STR_landed', 'B_avg_TOTAL_STR_att', 'B_avg_TOTAL_STR_landed', 'B_avg_opp_TOTAL_STR_att', 'B_avg_opp_TOTAL_STR_landed', 'B_avg_TD_att', 'B_avg_TD_landed', 'B_avg_opp_TD_att', 'B_avg_opp_TD_landed', 'B_avg_HEAD_att', 'B_avg_HEAD_landed', 'B_avg_opp_HEAD_att', 'B_avg_opp_HEAD_landed', 'B_avg_BODY_att', 'B_avg_BODY_landed', 'B_avg_opp_BODY_att', 'B_avg_opp_BODY_landed', 'B_avg_LEG_att', 'B_avg_LEG_landed', 'B_avg_opp_LEG_att', 'B_avg_opp_LEG_landed', 'B_avg_DISTANCE_att', 'B_avg_DISTANCE_landed', 'B_avg_opp_DISTANCE_att', 'B_avg_opp_DISTANCE_landed', 'B_avg_CLINCH_att', 'B_avg_CLINCH_landed', 'B_avg_opp_CLINCH_att', 'B_avg_opp_CLINCH_landed', 'B_avg_GROUND_att', 'B_avg_GROUND_landed', 'B_avg_opp_GROUND_att', 'B_avg_opp_

We see there are 109 columns with null values as shown in the list above.

We then redo this process with our fight statistics dictionary to see how many null values there are.

In [281]:
# Creating a dataframe to look at whether there are null values and how many
df2null = pd.DataFrame(df2pandas.isnull().sum())
df2null = df2null.rename(columns = {0:'count'})

In [282]:
# Looking at what the indices and their values are
indices2 = df2null.index
values2 = list(df2null.values)

They seem to be just the features. Thus, we can create a dictionary counting the number of null values for each feature.

In [283]:
nullvaluedict2 = dict(zip(indices2, values2))

In [284]:
nullvaluedict2

{'EVENT': array([0], dtype=int64),
 'BOUT': array([0], dtype=int64),
 'OUTCOME': array([0], dtype=int64),
 'WEIGHTCLASS': array([0], dtype=int64),
 'METHOD': array([0], dtype=int64),
 'ROUND': array([0], dtype=int64),
 'TIME': array([0], dtype=int64),
 'TIME FORMAT': array([0], dtype=int64),
 'REFEREE': array([32], dtype=int64),
 'DETAILS': array([71], dtype=int64),
 'URL': array([0], dtype=int64)}

We have a lot easier cleaning process with the fight statistics as there are no null values in any column besides 'Referee' and 'Details'.

In [285]:
z = []
for (x,y) in zip(nullvaluedict2.values(), nullvaluedict2):
    if x != 0:
        z.append(y)
print(z)
print(len(z))

['REFEREE', 'DETAILS']
2
